packages/net-stubbing/lib/server/state.ts

```diff
@@ -1,24 +1,60 @@
-import { noop } from 'lodash'
+import type { InterceptedRequest } from './intercepted-request'
 import type { NetStubbingState } from './types'
 
 export function state (): NetStubbingState {
   return {
     requests: {},
     routes: [],
     pendingEventHandlers: {},
-    reset () {
-      // clean up requests that are still pending
-      for (const requestId in this.requests) {
-        const { res } = this.requests[requestId]
+    async reset () {
+      await Promise.all(Object.values <InterceptedRequest>(this.requests).map((request) => {
+        const responseEvents = ['before:response', 'response:callback', 'response', 'network:error']
+        const inResponsePhase = responseEvents.includes(request.lastEvent!)
+        const { res } = request
 
-        res.removeAllListeners('finish')
-        res.removeAllListeners('error')
-        res.on('error', noop)
-        res.destroy()
-      }
+        if (res.destroyed) {
+          delete this.requests[request.id]
+
+          return Promise.resolve()
+        }
+
+        let shouldDestroyResponse = false
+
+        request.subscriptionsByRoute.forEach((subscriptionByRoute) => {
```

Comment:
Can maybe get some more legibility by replacing the `forEach` loops with more functional style `.some()` calls.

```
const shouldAwaitSubscription = (sub) => (
  sub.await
  && !sub.skip
  && (responseEvents.includes(sub.eventName) || !inResponsePhase)
)

const shouldDestroyResponse = request.subscriptionsByRoute.some((subs) => (
  !subs.immediateStaticResponse
  && subs.subscriptions.some(shouldAwaitSubscription)
})
```